<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN">
<html>
  <head>
    <meta content="text/html; charset=utf-8" http-equiv="content-type">
    <title>RNAstructure Command Line Help: partition</title>
    <link href="Stylesheet.css" rel="stylesheet" type="text/css">
  </head>
  <body>
    <div class="bodydiv">
      <table class="mainTable" border="0" cellpadding="0" cellspacing="0">
        <tbody>
          <tr class="headerRow">
            <td class="headerBox"><img src="icon.jpg" alt="RNAstructure logo"></td>
            <td class="headerLabel">
              <h2>RNAstructure Command Line Help<br>
                AlignmentPartition and AlignmentPartition-smp</h2>
            </td>
            <td class="headerBox2">
              <ul id="navigator">
                <li>
                  <h3 class="noMarginOrPad"><a href="index.html">Contents</a></h3>
                </li>
                <li>
                  <h3 class="noMarginOrPad"><a href="AlphabeticalIndex.html">Index</a></h3>
                </li>
              </ul>
            </td>
          </tr>
          <tr>
            <td class="mainContent" colspan="3">
              <p> AlignmentPartition is used to calculate the partition function for an alignment of homologous
                sequence, which can be used to predict base pair probabilities for consensus structures.
                <br>
              </p>
              <p>AlignmentPartition-smp is a parallel processing version for use on
                multi-core computers, built using OpenMP.</p>
              <a name="Usage" id="Usage"></a>
              <h3>USAGE: AlignmentPartition &lt;alignment file&gt; &lt;pfs file&gt; [options]</h3>
              <h3>OR: AlignmentPartition-smp &lt;alignment file&gt; &lt;pfs file&gt; [options]</h3>
              <h3 class="leftHeader">Required parameters:</h3>
              <table>
                <tbody>
                  <tr>
                    <td class="cell">&lt;alignment file&gt;</td>
                    <td>The name of a file containing an <a href="File_Formats.html#AFASTA">multiple sequence alignment</a> in FASTA format.<br>
                      Note that lowercase nucleotides are forced single-stranded
                      in structure prediction. </td>
                  </tr>
                  <tr>
                    <td class="cell">&lt;pfs file&gt;</td>
                    <td>The name of a binary partition function save file to
                      which output will be written.</td>
                  </tr>
                </tbody>
              </table>
              <h3 class="leftHeader">Options that do not require added values:</h3>
              <table>
                <tbody>
                  <tr>
                    <td class="cell">--disablecoax</td>
                    <td>Specify that coaxial stacking should not be used in the free energy calculation. This option uses a simpler energy function in exchange for a faster calculation.</td>
                  </tr>
                  <tr>
                    <td class="cell">-h, -H, --help</td>
                    <td>Display the usage details message.</td>
                  </tr>
                  <tr>
                    <td class="cell">-v --version</td>
                    <td>Display version and copyright information for this interface.<br></td>
                  </tr>
                </tbody>
              </table>
              <h3 class="leftHeader">Options that require added values:</h3>
              <table>
                <tbody>
                  <tr>
                    <td class="cell">-b -B --bpcutoff</td>
                    <td>Specify the base pairing cutoff.<br>
Default is 0.3</td>
                  </tr>
                  <tr>
                    <td class="cell">-l -L --loop</td>
                    <td>Specify a maximum internal/bulge loop size.<br>
Default is 40 unpaired nucleotides.</td>
                  </tr>
                </tbody>
              </table>
              <h3 class="leftHeader">Notes:</h3>
              <p>AlignmentPartition save files can be used with <a href="MaxExpect.html">MaxExpect</a> to estimate structures, <a href="ProbabilityPlot.html">ProbabilityPlot</a> to draw plots or provide text files with pair probabilities, or with <a href="draw.html">draw</a> to color annotate predicted structures. Stochastic sampling is not supported at this time.</p>
              <h3 class="leftHeader">Notes for smp:</h3>
              <p>partition-smp, by default, will use all available compute cores
                for processing. The number of cores used can be controlled by
                setting the OMP_NUM_THREADS environment variable.</p>
              <h3 class="leftHeader"><a name="Refs" id="Refs">References</a></h3>
              <ol class="noListTopMargin">
                <li class="gapListItem"> Reuter, J.S. and Mathews, D.H. <br>
                  "RNAstructure: software for RNA secondary structure prediction
                  and analysis." <br>
                  <i>BMC Bioinformatics</i>, 11:129. (2010). </li>
                <li class="gapListItem"> Mathews, D.H. <br>
                  "Using an RNA secondary structure partition function to
                  determine confidence in base pairs predicted by free energy
                  minimization." <br>
                  <i>RNA</i>, 10:1178-1190. (2004). </li>
                <li class="gapListItem"> McCaskill, J.S. <br>
                  "The equilibrium partition function and base pair
                  probabilities for RNA secondary structure." <br>
                  <i>Biopolymers</i>, 29:1105-1119. (1990). </li>
              </ol>
            </td>
          </tr>
          <tr>
            <td class="footer" colspan="3">
              <p>Visit <a href="http://rna.urmc.rochester.edu/RNAstructure.html">The
                  Mathews Lab RNAstructure Page</a> for updates and latest
                information.</p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>
</html>
